### ISSUE TITLE
calling mapArray in modules doesn't make the map implementation available to the module

### ISSUE THREAD

**Issue by djerius** [vectorgraphics/asymptote#400]
`mapArray` installs the generated map function somewhere that a module can't access. 

For example, `test.asy` below will compile as a standalone program, but not when imported as a  module:
```
% cat test.asy
mapArray("string","int");

int[] index_names( string[] expected, string[] got ) {
  return  map( new int( string check ) { return 1; }, got);
}
% asy  test.asy
% cat import.asy
import test;
% asy import.asy
  return  map( new int( string check ) { return 1; }, got);
             ^
test.asy: 4.14: no matching function 'map(int(string check), string[])'
% 
```

**Comment by johncbowman** [comment]
mapArray is implemented with eval (with embedded=true), so (quoting from  the documentation) it "will be evaluated at the top level of the current environment." So you need you need to call `mapArray("string","int");` from the top level and `include` (rather than` import`) the file `test.asy`.

**Comment by charlesstaats** [comment]
Telling people they have to `include` rather than `import` your module seems a bit unfortunate. Could an alternative version of `eval` be made to run the code in the current scope, for precisely this sort of issue? Alternatively, you could add a different version of `mapArray` specifically for use creating modules that writes a file to be imported, to be used something like this:

~~~asy
makemapArray("string","int");
import maparray_string_int;
~~~

**Comment by johncbowman** [comment]
I wish we could make `eval` run in the current scope but that is an open problem, unfortunately. Can you give me an example of how `makemapArray` would work?

**Comment by johncbowman** [comment]
Ok, I see what Charles is suggesting. I'm not sure about hard-coding a file name so for now I will leave this up to the user. But in the latest git version I've made it easier to do so by implementing `mapArray` in terms of `mapArrayString`.
The original example could then be written:

```
file f=output("maparray_string_int.asy");
write(f,mapArrayString("string","int"));
close(f);

import maparray_string_int;

int[] index_names( string[] expected, string[] got ) {
  return  map( new int( string check ) { return 1; }, got);
}

```

This example suggests a potential way to develop a more robust eval command: we should be able to support importing from a string rather than a file.

**Comment by djerius** [comment]
> Telling people they have to `include` rather than `import` your module seems a bit unfortunate. Could an alternative version of `eval` be made to run the code in the current scope, for precisely this sort of issue?

I wasn't really clear on why I raised the issure: I'm trying to write reuseable modules, and the compartmentalization provided by import (e.g. separate namespaces, no leaked symbols) is an important feature, so having to include rather than import to get a something breaks that paradigm.

**Comment by johncbowman** [comment]
Yes, import is preferable to include for those reasons but for now you can use the above workaround (which requires writing to and then importing a file), motivated by Charles' suggestion.

**Comment by johncbowman** [comment]
Followup: we now have a working prototype for templated imports and expect something similar to be included in the 2.88 release (after the upcoming 2.87 release).
